File: 074/cpp/main.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        int n = matrix[0].size();

        int i = -1;
        int j = m * n;
        while (i + 1 < j) {
            int k = i + (j - i) / 2;
            int t = matrix[k / n][k % n];
            if (t == target) {
                return true;
            } else if (t < target) {
                i = k;
            } else {
                j = k;
            }
        }

        if (i > -1 && matrix[i / n][i % n] == target) {
            return true;
        }
        if (j < m * n && matrix[j / n][j % n] == target) {
            return true;
        }
        return false;
    }
};

File: 074/rust/main.rs
impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        let m = matrix.len();
        let n = matrix[0].len();

        let mut i = 0;
        let mut j = m * n;
        while i + 1 < j {
            let k = i + (j - i) / 2;
            let t = matrix[k / n][k % n];
            if t == target {
                return true;
            }
            if t < target {
                i = k;
            } else {
                j = k;
            }
        }

        return matrix[i / n][i % n] == target;
    }
}
